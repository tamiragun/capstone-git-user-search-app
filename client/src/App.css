.App {
  text-align: center;
}

.main-body {
  display: flex;
  flex-direction: row;
  justify-content: space-around;
}

.results-list,
.repo-list {
  list-style: none;
  padding: 0;
}

.repo-list ul {
  border: dotted, 1px, lightgrey;
}

.single-user,
.single-repo,
.search-results {
  border: dotted 1px grey;
  padding: 20px;
}
