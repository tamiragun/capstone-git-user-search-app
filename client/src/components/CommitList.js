import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of an array of commit objects,
//and displays an ordered list of them.
export function CommitList(props) {
  //Extract the list of commits from the given props
  const commits = props.commits;

  //For every commit in the array of commits that was passed in as a prop,
  //display the message and the date. Also assign a unique key to each.
  const commitsList = commits.map((commit) => {
    return (
      <li key={commit.id} className="commit-bullet">
        {commit.message} <br></br>({commit.committed_date})
      </li>
    );
  });

  return <ol className="commits-list">{commitsList}</ol>;
}

//The prop is required, and should be an array of commit objects.
CommitList.propTypes = {
  commits: PropTypes.array.isRequired,
};
