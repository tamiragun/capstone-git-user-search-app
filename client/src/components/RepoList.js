import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of an array of repo objects,
//and displays an ordered list of them. It also takes a prop with the
//displayRepo function of the parent (SingleUser --> App) for when the
//button is clicked, and a third props which is the user object, which
//is used as an argument in that displayRepo function.
export function RepoList(props) {
  //Extract the list of repos from the given props
  const repos = props.repos;

  //Event handler which is triggered when the button is clicked, and in
  //turn invokes the displayRepo function which was passed down as a prop.
  const handleClick = (e) => {
    //As arguments, pass the source and username from the user prop, and
    //the value of the button that was clicked
    props.displayRepo(props.user.source, props.user.login, e.target.value);
  };

  //For every repo in the array of repos that was passed in as a prop,
  //display the name and assign a unique key to each.
  const reposList = repos.map((repo) => {
    return (
      <li key={repo.id}>
        <div className="repo-bullet">
          <span>{repo.name}</span>
          <button
            //Assign two values to this, as both are needed by the server API
            //This results in a comma separated string which will be handled
            //by the function it is passed to
            value={[repo.name, repo.id]}
            onClick={handleClick}
            className="select-button view-repo-button"
          >
            View this repo
          </button>
        </div>
      </li>
    );
  });

  return <ol className="repo-list">{reposList}</ol>;
}

//The props are required, and should be an array of commit objects, an object,
//and a function.
RepoList.propTypes = {
  repos: PropTypes.array.isRequired,
  user: PropTypes.object.isRequired,
  displayRepo: PropTypes.func.isRequired,
};
