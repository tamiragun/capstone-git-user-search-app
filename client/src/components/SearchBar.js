import React from "react";
import PropTypes from "prop-types";

//This component takes a prop in the form of a search function, and displays
//a form. When input is entered, this component updates its own state and when
//the form is submitted, is passes the info to the search function of the
//parent (App).
export class SearchBar extends React.Component {
  //Set the state to have the search term as an empty string
  constructor(props) {
    super(props);
    this.state = { searchString: "" };

    //Bind the two functions defined below
    this.handleUserInput = this.handleUserInput.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  //Each time something is typed in the input field, update the
  //state to reflect the latest value.
  handleUserInput(e) {
    this.setState({ searchString: e.target.value });
  }

  //Once the form is submitted, execute the search
  handleSubmit(e) {
    //Prevent the browser from refreshing
    e.preventDefault();
    //Call the function passed as a prop by the parent (App)
    //And pass it the latest state of the searchString.
    this.props.handleSearch(this.state.searchString);
    //Clear the input field by setting the state back to an empty string
    this.setState({ searchString: "" });
  }

  render() {
    return (
      <div>
        {/*Upon form submit, call the handler function defined above */}
        <form onSubmit={this.handleSubmit}>
          <input
            type="text"
            //Set the value to whatever the current state is, to control this input field
            value={this.state.searchString}
            //Every time its content changes, call the change handler to update the state
            onChange={this.handleUserInput}
            className="search-input"
            required
          ></input>
          <button type="submit" className="select-button">
            Search
          </button>
        </form>
      </div>
    );
  }
}

//The prop is required, and should be a function.
SearchBar.propTypes = {
  handleSearch: PropTypes.func.isRequired,
};
