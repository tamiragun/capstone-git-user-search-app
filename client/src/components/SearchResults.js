import React from "react";

export function SearchResults(props) {
  const results = props.results;
  const handleClick = (e) => {
    props.displayUser(e.target.value);
  };
  const resultsList = results.map((result) => {
    return (
      <li key={result.id} className="bullet">
        <p>
          <strong>Name: </strong>
          {result.login} <br></br>
        </p>

        <button value={result.login} onClick={handleClick}>
          View this user
        </button>
      </li>
    );
  });

  return (
    <div className="search-results">
      <h3>Search results:</h3>
      <ul className="results-list">{resultsList}</ul>
    </div>
  );
}
