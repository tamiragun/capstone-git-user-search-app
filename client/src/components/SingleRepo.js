import React from "react";
import { CommitList } from "./CommitList";
import PropTypes from "prop-types";
import { getData } from "../utils/getData.js";

//This component takes a prop in the form of a repo object, and displays
//its details. It also takes a prop with the backToUser function of the
//parent (App) for when the button is clicked
export function SingleRepo(props) {
  //commits will determine whether the commits are displayed vs a loading
  //placeholder, and contain the necessary top 5 commits info
  const [commits, setCommits] = React.useState(null);

  //As soon as the component is rendered, this function will run and fetch
  //the list of commits to display.
  React.useEffect(() => {
    //Nest an async function within useEffect, because useEffect is synchronous
    const fetchData = async () => {
      //Construct the server endpoint url using the props
      const url = `/api/commits?source=${props.repo.source}&user=${props.repo.owner}&reponame=${props.repo.name}&repoid=${props.repo.id}`;
      // Fetch the data
      const result = await getData(url);
      //Update the state to contain the array of commits
      setCommits(result);
    };
    //Call the async function to execute the fetch
    fetchData();
    //Run this effect every time the repo info changes
  }, [props.repo]);

  return (
    <div className="single-repo">
      <h2>{props.repo.name}</h2>
      <div className="TBC">
        <p>
          {props.repo.description}
          <br></br>
          <br></br>
          <strong>First created: </strong>
          {props.repo.created_at}
          <br></br>
          <strong>Last updated: </strong>
          {props.repo.updated_at}
          <br></br>
        </p>
        <h3>Latest 5 commits:</h3>
        {/*If the commits fetch is till pending, display a placeholder */}
        {!commits ? (
          <div className="placeholder">Loading...</div>
        ) : /*If the commits fetch yields zero results, display a message */
        commits.length === 0 ? (
          <div className="placeholder">No commits yet</div>
        ) : (
          /*If the commits fetch yields results, display the commits. Pass 
          on the list of commits from the state. */
          <CommitList commits={commits}></CommitList>
        )}
      </div>
      <div className="TBC">
        {/*No need for a separate event handler, this invokes the function 
          that was passed down as a prop */}
        <button onClick={props.backToUser} className="back-button">
          Go back to user profile
        </button>
      </div>
    </div>
  );
}

//The props are required, and should be a repo object and a function.
SingleRepo.propTypes = {
  repo: PropTypes.object.isRequired,
  backToUser: PropTypes.func.isRequired,
};
