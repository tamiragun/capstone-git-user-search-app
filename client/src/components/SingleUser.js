import React from "react";
import { RepoList } from "./RepoList";
const fetch = require("node-fetch");

export function SingleUser(props) {
  const [repos, setRepos] = React.useState(null);

  React.useEffect(async () => {
    const url = `http://localhost:3001/api/repos?source=${props.user.source}&user=${props.user.login}`;
    try {
      const response = await fetch(url);
      //console.log("Response: ", response, " Type: ", typeof response);
      const jsonResponse = await response.json();
      // console.log(
      //   "jsonResponse: ",
      //   jsonResponse,
      //   " Type: ",
      //   typeof jsonResponse
      // );
      setRepos(jsonResponse);
    } catch (err) {
      console.log(err);
    }
  }, [props.user.login]);

  const handleClick = (e) => {
    props.goBack();
  };

  return (
    <div className="single-user">
      <h3>Selected user</h3>
      <div className="TBC">
        <p>
          <strong>Name: </strong>
          {props.user.login}
          <br></br>
          <strong>Url: </strong>
          {props.user.url}
          <br></br>
        </p>
        {!repos ? (
          "Loading"
        ) : (
          <RepoList
            user={props.user.login}
            repos={repos}
            displayRepo={props.displayRepo}
          ></RepoList>
        )}
      </div>
      <div className="TBC">
        <button onClick={props.goBack}>Go back</button>
      </div>
    </div>
  );
}
