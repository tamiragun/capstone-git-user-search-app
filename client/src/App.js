import React from "react";
import { SearchBar } from "./components/SearchBar";
import { SearchResults } from "./components/SearchResults";
import { SingleUser } from "./components/SingleUser";
import { SingleRepo } from "./components/SingleRepo";
import "./App.css";

function App() {
  //Declare the various state hooks

  //User will determine whether a single user is displayed, and contain the
  //necessary user info
  const [user, setUser] = React.useState(null);
  //Repo will determine whether a single repo is displayed, and contain the
  //necessary repo info
  const [repo, setRepo] = React.useState(null);
  //Results will determine whether the results are displayed vs a loading
  //placeholder, and contain the necessary results info
  const [results, setResults] = React.useState(null);
  //ShowResults will determine whether the results (loading placeholder or list)
  //are displayed or not
  const [showResults, setShowResults] = React.useState(false);

  //When a search is triggered by the child component (SearchBar), call the
  //server to get the search results for that particular search term
  const handleSearch = async (term) => {
    //First, set the other states to null so that the previous cards disappear
    //from the screen.
    setUser(null);
    setRepo(null);
    setResults(null);
    //Allow the app to render the search results (which will first show 'loading'
    //until the data is received, then it will display the data)
    setShowResults(true);
    //Construct the server endpoint url using the argument passed into the function
    const url = `/api/search?term=${term}`;
    try {
      //Call the endpoint
      const response = await fetch(url);
      const jsonResponse = await response.json();
      //Update the state to contain the array of search results
      setResults(jsonResponse);
      //If the request is unsuccessful, print the error message to the console
    } catch (err) {
      console.log(err);
    }
  };

  //When the "Show this user" button is triggered by the child component (SearchResults), call the
  //server to get the user results for that particular user
  const displayUser = async (input) => {
    //First, set the other states to null so that the previous cards disappear
    //from the screen.
    setUser(null);
    setRepo(null);
    //The argument passed into the function is a string separated by a comma.
    //To access each element separately, split the string so it becomes an array
    //The first element should be the source and the second should be the username
    const inputArray = input.split(",");
    const source = inputArray[0];
    const login = inputArray[1];
    //Construct the server endpoint url
    const url = `/api/user?source=${source}&user=${login}`;
    try {
      //Call the endpoint
      const response = await fetch(url);
      const jsonResponse = await response.json();
      //Update the state to contain the returned user object
      setUser(jsonResponse);
      //If the request is unsuccessful, print the error message to the console
    } catch (err) {
      console.log(err);
    }
  };

  //When the "Show this repo" button is triggered by the child component (SingleUser), call the
  //server to get the repo results for that particular user
  const displayRepo = async (source, user, repo) => {
    //First, set the repo state to null so that the previous card disappears
    //from the screen.
    setRepo(null);
    //The argument passed into the function is a string separated by a comma.
    //To access each element separately, split the string so it becomes an array
    //The first element should be the name and the second should be the id
    const repoData = repo.split(",");
    const repoName = repoData[0];
    const repoId = repoData[1];
    //Construct the server endpoint url
    const url = `/api/repo?source=${source}&user=${user}&reponame=${repoName}&repoid=${repoId}`;
    try {
      //Call the endpoint
      const response = await fetch(url);
      const jsonResponse = await response.json();
      //Update the state to contain the returned repo object
      setRepo(jsonResponse);
      //If the request is unsuccessful, print the error message to the console
    } catch (err) {
      console.log(err);
    }
  };

  //When the "Go back to search results" button is clicked on the child component
  //(SingleUser), the user and repo status should be reset to null so that they
  //vanish from the screen and only the search results are shown
  const backToSearch = () => {
    setUser(null);
    setRepo(null);
  };

  //When the "Go back to user" button is clicked on the child component
  //(SingleRepo), the repo status should be reset to null so that it
  //vanishes from the screen and only the search results and user are shown
  const backToUser = () => {
    setRepo(null);
  };

  //When the "Search again" button is clicked on the child component
  //(SearchResults), the results, user and repo status should be reset to null
  //so that they vanish from the screen. the showResults state should be set to
  //false so that only the search bar is shown
  const backToHome = () => {
    setUser(null);
    setRepo(null);
    setResults(null);
    setShowResults(false);
  };

  return (
    <div className="App">
      <header className="App-header">
        <h1>Search users across Github and Gitlab</h1>
      </header>
      <div className="main-body">
        <SearchBar handleSearch={handleSearch}></SearchBar>
        {/*Search results are displayed if the state showSearch is set to true*/}
        {showResults && (
          <div className="search-results">
            <h2>Search results:</h2>
            {/*Search results show a placeholder if the results are still null*/}
            {!results ? (
              <div className="placeholder">Loading...</div>
            ) : /*Search results show a message if the results are empty*/
            results.length === 0 ? (
              <div className="placeholder">No results on Github or Gitlab.</div>
            ) : (
              /*Search results show the results once the data is in*/
              <SearchResults
                displayUser={displayUser}
                results={results}
              ></SearchResults>
            )}
            <button onClick={backToHome} className="back-button">
              Search again
            </button>
          </div>
        )}
        <div className="single-cards">
          {/*User card is displayed only if the state has a user object set*/}
          {user && (
            <SingleUser
              backToSearch={backToSearch}
              user={user}
              displayRepo={displayRepo}
            ></SingleUser>
          )}
          {/*Repo card is displayed only if the state has a repo object set*/}
          {repo && (
            <SingleRepo backToUser={backToUser} repo={repo}></SingleRepo>
          )}
        </div>
      </div>
    </div>
  );
}

export default App;
