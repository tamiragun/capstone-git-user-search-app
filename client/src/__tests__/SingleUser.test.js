import React from "react";
import { SingleUser } from "../components/SingleUser";
import renderer from "react-test-renderer";
import App from "../app.js";

test("SingleUser is shown", () => {
  const user = {
    id: 37237140,
    name: "Tamira",
    login: "tamiragun",
    avatar: "https://avatars.githubusercontent.com/u/37237140?v=4",
    url: "https://api.github.com/users/tamiragun",
    source: "Github",
  };
  const tree = renderer
    .create(
      <SingleUser
        user={user}
        backToSearch={function backToSearch() {}}
        displayRepo={function displayRepo() {}}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// test("displayUser fetches data correctly", () => {
//   return App.displayUser("Github,tamiragun").then((data) => {
//     expect(data).toEqual([
//       {
//         id: 37237140,
//         name: "Tamira",
//         login: "tamiragun",
//         avatar: "https://avatars.githubusercontent.com/u/37237140?v=4",
//         url: "https://api.github.com/users/tamiragun",
//         source: "Github",
//       },
//     ]);
//   });
// });
