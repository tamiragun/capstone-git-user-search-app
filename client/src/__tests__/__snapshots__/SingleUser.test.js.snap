// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleUser is shown 1`] = `
<div
  className="single-user"
>
  <h2>
    tamiragun
  </h2>
  <div
    className="TBC"
  >
    <img
      alt="user avatar"
      src="https://avatars.githubusercontent.com/u/37237140?v=4"
    />
    <br />
    <a
      href="https://api.github.com/users/tamiragun"
    >
      https://api.github.com/users/tamiragun
    </a>
    <h3>
      Latest 5 repositories:
    </h3>
    <div
      className="placeholder"
    >
      Loading...
    </div>
  </div>
  <div
    className="TBC"
  >
    <button
      className="back-button"
      onClick={[Function]}
    >
      Go back to search results
    </button>
  </div>
</div>
`;
